### PR TITLE
[NativeAOT][8.0] Make casting logic closer to CoreCLR

### LINKED ISSUE #84464: [NativeAOT] Possible perf improvements in casting
CoreClr appears to have faster casting helpers than NativeAOT. https://github.com/dotnet/runtime/pull/84430#issuecomment-1499671856

That means the NativeAOT helpers could likely be improved, since conceptually they should be applying similar logic.
It is worth compare the code of the helpers and related typesystem bits to see what can be copied/unified from whichever is a faster implementation.

### PR DESCRIPTION
**Pull request by VSadov**
Reduced version of https://github.com/dotnet/runtime/pull/89548  (no JIT changes)

Fixes: https://github.com/dotnet/runtime/issues/84464

The actual changes are not as big here as might seem. For the most part this is a refactoring of existing code to have a shape closer to CoreCLR cast helpers, so that similar patterns could be used - in a few places where that has not been done already in earlier changes.  
For example cases like `CheckCastAny` - could start with a cache lookup, since uncached code path can be complex and thus relatively slow. (also addresses some old TODOs in this area)

### CODE REVIEW

File: src/tests/baseservices/typeequivalence/simple/Simple.cs [hunk 1]
```diff
@@ -268,6 +274,17 @@ private static void LoadInvalidType()
         Console.WriteLine($"-- {typeof(ValueTypeWithInstanceMethod).Name}");
     }
 
+    private static void TestCastsOptimizations()
```
Comment (jkotas) [inline src/tests/baseservices/typeequivalence/simple/Simple.cs:277]: I thought that this test needs a fix in the JIT and JIT/EE interface to pass.

Comment (VSadov) [reply src/tests/baseservices/typeequivalence/simple/Simple.cs:277]: If we do not optimize isinst that classified as ANY, the test should pass. 

The troubles start when array casts are classified as ANY, then we need to optimize them, but then ANY brings other subtle cases (equivalence, nullable) that are not optimizable.

Comment (jkotas) [reply src/tests/baseservices/typeequivalence/simple/Simple.cs:277]: Ok, I see.

**Review by jkotas — APPROVED** [review]
Thank you!

**Comment by VSadov** [comment]
Thanks!!

**Pull request merged**